core/src/transaction.rs

```diff
@@ -204,41 +204,38 @@ impl OutPoint {
 pub struct CellInput {
     pub previous_output: OutPoint,
     pub since: u64,
-    // Depends on whether the operation is Transform or Destroy, this is the proof to transform
-    // lock or destroy lock.
-    pub args: Vec<Bytes>,
+    // Only used for cellbase input, normal input doesn't need this field
+    pub block_number: BlockNumber,
```

Comment:
Is this a temporary workaround?

Reply:
Nope this is a solution I'm proposing here. Since args is absent we need a way to tell the difference between each cellbase input.

Reply:
We can use cellbase input's witness.

Reply:
The problem with witness, is that it won't be included when calculating TX hash. So if a miner creates 2 cellbase transactions within the epoch, the hash would be identical, hence there's no way to tell one from the other.

Reply:
Another choice: can we require cellbase tx to have a `dep` refer to its own header?

Reply:
Isn't that a circular dependency? A header would include transaction root, which needs the hashes of all transactions(including cellbase tx), and the hash of the cellbase tx cannot be calculated until we fill in the header hash in `dep`.

Reply:
That's a circular dependency. Block hash depends on cellbase hash, cellbase requires block hash

Reply:
But we can let it depends on parent block header

Reply:
What about use `block_hash` to reference parent block header? We can keep `cell` in `OutPoint` as null to represent a cellbase OutPoint, but `block_hash` can be used (it's designed to reference a block header anyway).

Reply:
It still smells a bit strange, what about force the first output use block_number as data?

Reply:
After some thought I do feel it's better if we leverage the data field of the first output in the cellbase transaction. We can add a verification rule so that the first 4 bytes of the first output's data should store the block number. This still leaves space if the miner wants to encode more message. On the other hand, this avoid one more quirky rule in OutPoint handling.

Thoughts? @doitian @janx

Reply:
The first-class inflation will stop when block issuance is smaller than cellbase output capacity requirement. Use cellbase output to store extra data has impact on first-class inflation.

Reply:
Should be ok given our current second-class issuance setting.

Reply:
Another side of the point is: if the issuance is not enough for a cell, there shouldn't be a cell base tx, correct?

Reply:
I think we can just assume that won't happen to simplify the code.

Reply:
This is now updated.

Reply:
Another approach (which I think is better) is to reuse `since` field to store block_number, and we can make sure it works without addition assumption, leave the first output data to other usages.

Reply:
I'm not sure the logic of validating `since` alone is enough here. We will still need to verify if the `since` field stores the correct block number. Otherwise an attacker or a careless user might just skip filling in `since`, still resulting in cellbase transactions that are identical.

And if we will need the logic to verify block number anyway, I think putting it in data part is better for separation of concerns. In addition, it's not that the data field for this output is not usable, we are only reserving the first 8 bytes for block number. The miner is still free to put whatever data he/she wants after the initial 8 bytes.

Reply:
I actually think use `since` for block number is better, as `since` is designed for block number. Another not-so-important point to avoid use cellbase data here is it also increases occupied capacity which affects second-issuance to miner.

A rule to make sure cellbase input filled `since` with current block number is still needed.

Reply:
Yes, we will need the same logic to check the block number. 

The reason I like the `since` idea is we do not need to rely on output data in our consensus, the initial 8 bytes implies that the first output cannot be constrained under some special types, which limit the usage, for example, a potential usage, can a miner put outputs into NervosDAO directly?

Just some thoughts, I do not strongly insist on this.

Reply:
For note, put `block_number` to `since` is against the original `since` rule, the default `0x0000` flags require tip number at least equals to `since`, which is always fails because in our case the block number is `tip + 1`.

I think we can break the `since` rule for cell base tx, because in semantics `since` is meaningless for cell base tx.

Reply:
I have mixed feelings on this: on the one hand, I actually agree that using since is better than output data part due to the output usage point mentioned above, on the other hand, I don't think it's ideal to break since rule just for cell base, we've had enough special rules, and adding one here is only gonna add to the complexity.

One thing I'm thinking is: can we expand since rule with another flag to require that the block number put in since field must equal current block number? This way we won't break since rule, we are just adding a feature here.

Reply:
My mistake, actually, it does not break the `since` rule, when validating a block we should use block_number as tip_number.

Reply:
I've updated the code to keep block number in `since` field, I can confirm that original since rule works here.

Reply:
upvote `since`, `cellbase` with block number is also helpful for reward tx proposer.